File: Code Python/05. Prediction/05.3 Compressed Sensing/SignalReconstructionWithCompressedSensing.py
import numpy as np
import matplotlib.pyplot as plt
import cvxpy as cvx

# Create Signal

dt = .001
t = np.arange(0, 1, dt)

f = np.sin(2*np.pi*10*t)+np.sin(2*np.pi*70*t) + np.sin(2*np.pi*160*t) + np.cos(2*np.pi*120*t)

plt.figure()
plt.subplot(2, 1, 1)
plt.plot(t, f, linewidth=2)
plt.title('Original signal')

FC = np.fft.fft(f)
P = np.abs(FC)/len(t)

freq = 1/(dt * len(t))*(np.arange(1, len(t)+1, 1))
plt.subplot(2, 1, 2)
plt.plot(freq, P, linewidth=2)
plt.xlim([1, 500])
plt.xlabel('frequencies in Hz')
plt.ylabel('Power of frequencies')
plt.title('Powerspectrum')
plt.show()

# Downsample Data
s = 55
red = np.random.randint(1000, size=s)
X = f[red]

plt.figure()
plt.plot(t,f,linewidth=2)
plt.plot(red/1001,X,'xr',linewidth=4)
plt.legend(['Original signal','Random samples'])
plt.title('Measurement samples')
plt.show()

# Create PSI and determine L1 minimization
DFT = np.conj(np.fft.fft(np.eye(1000))/1000)
PSI = DFT[red, :]

xL1 = cvx.Variable(1000, complex='True')
obj = cvx.Minimize(cvx.norm(xL1,1))
constraints = [cvx.sum_squares(PSI@xL1-X) <= 1]
prob = cvx.Problem(obj, constraints)
prob.solve()

xL1 = xL1.value
print(xL1)
xL2 = np.linalg.pinv(PSI) @ X

sigl1 = np.fft.ifft(xL1)
sigl2 = np.fft.ifft(xL2)

Pl1 = np.abs(xL1)/len(t)
Pl2 = np.abs(xL2)/len(t)

plt.figure(figsize=(20,10))
plt.subplot(2,2,1)
plt.title('Compressed Sensing with L1 norm')
plt.plot(t,f,linewidth=2)
plt.plot(t,np.real(sigl1),'r',linewidth=2)
plt.plot(red/1001,X,'xk',linewidth=2)
plt.legend(['Original signal','Reconstructed signal','Measurement samples'])
plt.subplot(2,2,3)
plt.plot(freq,Pl1,linewidth=2)
plt.xlabel('frequencies in Hz')
plt.ylabel('Power of frequencies')
plt.xlim([1, 500])
plt.title('Powerspectrum')

plt.subplot(2,2,2)
plt.title('Least square approximation')
plt.plot(t,f,linewidth=2)
plt.plot(t,np.real(sigl2),'r',linewidth=2)
plt.plot(red/1001,X,'xk',linewidth=2)
plt.legend(['Original signal','Reconstructed signal','Measurement samples'])
plt.subplot(2,2,4)
plt.plot(freq,Pl2,linewidth=2)
plt.xlabel('frequencies in Hz')
plt.ylabel('Power of frequencies')
plt.xlim([1, 500])
plt.title('Powerspectrum')

plt.show()

File: Code Matlab/05. Prediction/05.3 Compressed Sensing/SignalReconstructionWithCompressedSensing.m
clc, clear, close all
%% Signal

dt = 0.001;
t = (0:dt:1);

f = sin(2*pi*10*t)+sin(2*pi*70*t) +sin(2*pi*160*t) + cos(2*pi*120*t);

figure('Name','Signal and Power spectrum');
subplot(2,1,1)
plot(t,f,'LineWidth',2);
title('Original signal')

%% Fast Fourier Transformation 

% FC are the Fourier Coeficients
FC = fft(f,length(t));

% Power of different frequencies
P = abs(FC)/length(t);

freq = 1/(dt*length(t))*(1:length(t));

subplot(2,1,2)
hold on
plot(freq,P,'LineWidth',2);
xlim([1, 500]);
xlabel('frequencies in Hz')
ylabel('Power of frequencies')
title('Powerspectrum')

%% Downsample data

s = 55;
red = round(rand(s,1)*1001);
R = f(red);

figure('Name','Downsampled Data')
hold on
plot(t,f,'LineWidth',2)
plot(red/1001,R,'xr','LineWidth',4)
legend('Original signal','Random samples')
title('Measurement samples')


%% Create DFT and PSI determine x

DCT = dftmtx(1001);
PSI = conj(DCT(red,:)./length(t));

cvx_begin;
    variable xL1(1001) complex; % sparse vector of coefficients 
    minimize( 10*norm(xL1,1) );
    subject to
        PSI*xL1 == R';
cvx_end;

xL2 = pinv(PSI)*R';

%% Plot 

frecL1 = ifft(xL1);
frecL2 = ifft(xL2);

Pl1 = abs(xL1)/length(t);
Pl2 = abs(xL2)/length(t);

c = [0 0.4470 0.7410];

figure()
subplot(2,2,1)
title('Compressed Sensing with L1 norm')
hold on
plot(t,f,'LineWidth',2);
plot(t,real(frecL1),'r','LineWidth',2);
plot(red/1001,R,'xk','LineWidth',2)
legend('Original signal','Reconstructed signal','Measurement samples')
subplot(2,2,3)
plot(freq,Pl1,'LineWidth',2);
xlabel('frequencies in Hz')
ylabel('Power of frequencies')
xlim([1, 500])
title('Powerspectrum')

subplot(2,2,2)
title('Least square approximation')
hold on
plot(t,f,'LineWidth',2);
plot(t,real(frecL2),'r','LineWidth',2);
plot(red/1001,R,'xk','LineWidth',2)
legend('Original signal','Reconstructed signal','Measurement samples')
subplot(2,2,4)
plot(freq,Pl2,'LineWidth',2);
xlabel('frequencies in Hz')
ylabel('Power of frequencies')
xlim([1, 500])
title('Powerspectrum')

%% Optimization of sparse solution with least squares

ind = Pl1 > 0.1;

xoptL1 = pinv(PSI*diag(ind))*R';

figure('Name','Optimized L1 with least square')
subplot(2,1,1)
title('Least square optimized sparse L_1 norm solution')
hold on
plot(f,'LineWidth',2)
plot(real(ifft(xoptL1)),'r','LineWidth',2)
xlim([1, 1000])
legend('Original signal','Reconstructed signal')
subplot(2,1,2)
plot(abs(xoptL1)/length(t),'LineWidth',2)
xlabel('frequencies in Hz')
ylabel('Power of frequencies')
xlim([1, 500])
title('Powerspectrum')
